File: AIペアプログラミング実践編_ソースコード/practical-python/session-4/exercise/inventory/services/product_service.py
"""商品サービス"""
from inventory.models.product import Product

from .dto import ProductInput
from .product_validator import ProductValidator


class ProductService:
    """商品のビジネスロジック"""

    def __init__(self, repository=None):
        self.repository = repository
        self.validator = ProductValidator()

    def create(self, input_data: ProductInput) -> Product:
        """商品を登録"""
        # バリデーション
        result = self.validator.validate(input_data)
        if not result.is_valid:
            raise ValueError(", ".join(result.errors))

        # 商品を作成
        product = Product(
            name=input_data.name,
            category=input_data.category,
            price=input_data.price
        )

        # 保存
        return self.repository.save(product)

    def find_all(self):
        """全商品を取得"""
        return self.repository.find_all()

    def find_by_id(self, product_id: int) -> Product:
        """ID で商品を取得"""
        product = self.repository.find_by_id(product_id)
        if product is None:
            raise ValueError(f"商品が見つかりません: {product_id}")
        return product

File: AIペアプログラミング実践編_ソースコード/practical-csharp/session-3/Ans/Inventory.Api/Services/ProductService.cs
namespace Inventory.Api.Services;

using Inventory.Api.Dtos;
using Inventory.Api.Entities;
using Inventory.Api.Repositories;

public class ProductService
{
    private readonly IProductRepository _repository;
    private readonly ProductValidator _validator;

    public ProductService(IProductRepository repository)
    {
        _repository = repository;
        _validator = new ProductValidator();
    }

    public Product Create(ProductInput input)
    {
        var result = _validator.Validate(input);
        if (!result.IsValid)
        {
            throw new ArgumentException(string.Join(", ", result.Errors));
        }

        var product = new Product
        {
            Name = input.Name,
            Category = input.Category,
            Price = input.Price,
            CreatedAt = DateTime.Now,
            UpdatedAt = DateTime.Now
        };

        return _repository.Save(product);
    }

    public List<Product> FindAll()
    {
        return _repository.FindAll();
    }

    public Product? FindById(int id)
    {
        return _repository.FindById(id);
    }

    public List<Product> Search(string? name, ProductCategory? category)
    {
        return _repository.Search(name, category);
    }
}
